service/matching/version_sets.go

```diff
@@ -280,14 +276,23 @@ func makeVersionInSetDefault(data *persistencespb.VersioningData, setIx, version
 	}
 }
 
-// Requires: data is not nil, caps is not nil
+// Requires: caps is not nil
 func lookupVersionSetForPoll(data *persistencespb.VersioningData, caps *commonpb.WorkerVersionCapabilities) (string, error) {
-	// for poll, only the latest version in the compatible set can get tasks
-	// find the version set that this worker is in
+	// For poll, only the latest version in the compatible set can get tasks.
+	// Find the version set that this worker is in.
+	// Note data may be nil here, findVersion will return -1 then.
 	setIdx, _ := findVersion(data, caps.BuildId)
 	if setIdx < 0 {
-		// TODO: consider making an ephemeral set so we can match even if replication fails
-		return "", errBuildIDNotFound
+		// A poller is using a build ID but we don't know about that build ID. This can happen
+		// in a replication scenario if pollers are running on the passive side before the data
+		// has been replicated. Instead of rejecting, we can guess a set id based on the build
+		// ID. If the build ID was the first in its set on the other side, then our guess is
+		// right and things will work out. If not, then we'll guess wrong, but when the
+		// versioning data replicates, we'll redirect the poll to the correct set id.
+		// In the meantime (e.g. during an ungraceful failover) we can at least match tasks
+		// using the exact same build ID.
+		guessedSetID := hashBuildID(caps.BuildId)
```

Comment:
We should have some metric for this

Reply:
I like a metric in theory but... we can't put the build id itself in the metric for cardinality reasons, so a metric actually actionable? a log seems better...

Reply:
Hmmm... task queue tag should be fine and I'd also add a log to know which build ID was guessed.

Reply:
added TODOs